lib/sites-base.js

```diff
@@ -1,214 +1,308 @@
+          help: 'like http://localhost:3000',
+          type: 'url'
+        },
+        {
+          name: 'stagingBaseUrl',
+          label: 'Staging Base URL',
+          help: 'like http://project.staging.org',
+          type: 'url'
+        },
+        {
+          name: 'prodBaseUrl',
+          label: 'Production Base URL',
+          help: 'like https://myproject.com',
+          type: 'url'
+        }
+      ].concat(options.addFields || []);
 
-    async function body() {
-      try {
-        await copyContent(req, piece, options);
-      } catch (e) {
-        return callback(e);
-      }
-      return callback(null);
+      options.arrangeFields = [
+        {
+          name: 'urls',
+          label: 'URLs',
+          fields: [
+            'hostnamesArray',
+            'devBaseUrl',
+            'stagingBaseUrl',
+            'prodBaseUrl'
+          ]
+        }
+      ].concat(options.arrangeFields || []);
     }
+  },
 
-    async function copyContent(req, piece, options) {
-
-      if (!piece.copyOfId) {
-        return;
-      }
-
-      let from, to;
-
-      // copy the database and the attachments
-      await copyDatabase();
-      from = await getFrom();
-      to = await getTo();
-      await purgeTrash();
-      if (self.options.copy && self.options.copy.hardlink) {
-        await hardlinkAttachments();
-      } else {
-        await copyAttachments();
-      }
-
-      async function copyDatabase() {
-        const shortNamePrefix = self.apos.shortName.replace(/dashboard$/, '');
-        // http://stackoverflow.com/questions/36403749/how-can-i-execute-db-copydatabase-through-nodejss-mongodb-native-driver
-        const command = {
-          copydb: 1, fromdb: shortNamePrefix + piece.copyOfId, todb: shortNamePrefix + piece._id
-        };
-        const admin = self.apos.db.admin();
-        return await admin.command(command);
-      }
-
-      async function getFrom() {
-        return await self.apos.options.multisite.getSiteApos(piece.copyOfId);
-      }
-
-      async function getTo() {
-        return await self.apos.options.multisite.getSiteApos(piece._id);
+  construct: function(self, options) {
+    self.beforeSave = function(req, doc, options, callback) {
+      if ((!doc._id) && (!doc.copyOfId)) {
+        if (!doc.adminPassword) {
+          self.apos.notify(req, 'You must fill out the admin password field when creating a new site.', { type: 'error' });
+          return callback('adminPassword.required');
+        }
       }
-
-      async function purgeTrash() {
-        await to.docs.db.remove({ trash: true });
-        await to.attachments.db.remove({ trash: true });
+      // Don't let an admin password wind up in the db in cleartext
+      req.newAdminPassword = doc.adminPassword;
+      delete doc.adminPassword;
+  
+      doc.hostnames = _.map(doc.hostnamesArray || [], function(value) {
+        return value.hostname.toLowerCase().trim();
+      });
+  
```

Comment:
Where is the logic to populate devBaseUrl, stagingBaseUrl, and prodBaseUrl from the baseUrlDomains property and shortName?

Reply:
On line 95
`doc[key+'BaseUrl'] = shortname + '.' + self.options.baseUrlDomains[key].toLowerCase().trim()`

I started by writing 
`if (self.options.baseUrlDomains.dev) { ... } else if..` but I ended up iterating in the object `self.options.baseUrlDomains` and checking the key (that is to say 'dev', 'staging' or 'prod'). Would you prefer I add a comment or I explicitly add `if`s conditions?